src/mlpack/methods/rbm/rbm_impl.hpp

```diff
@@ -0,0 +1,206 @@
+    RBMPolicy rbmPolicy,
+    const size_t numSteps,
+    const size_t mSteps,
+    const bool useMonitoringCost,
+    const bool persistence):
+    rbmPolicy(rbmPolicy),
+    initializeRule(initializeRule),
+    numSteps(numSteps),
+    mSteps(mSteps),
+    useMonitoringCost(useMonitoringCost),
+    persistence(persistence),
+    reset(false)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>
+    ::Reset()
+{
+  size_t weight = 0;
+  weight+= rbmPolicy.Parameters().n_elem;
+  positiveGradient.set_size(weight, 1);
+  negativeGradient.set_size(weight, 1);
+  parameter.set_size(weight, 1);
+  parameter.zeros();
+  positiveGradient.zeros();
+  negativeGradient.zeros();
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+  rbmPolicy.Parameters() = arma::mat(parameter.memptr(), weight, 1, false,
+      false);
+  rbmPolicy.PositiveGradient() = arma::mat(positiveGradient.memptr(), weight, 1,
+      false, false);
+  rbmPolicy.NegativeGradient() = arma::mat(negativeGradient.memptr(), weight, 1,
+      false, false);
+
+  rbmPolicy.Reset();
+  gibbsTemporary.set_size(rbmPolicy.VisibleLayer().InSize(), 1);
+  negativeSamples.set_size(rbmPolicy.VisibleLayer().InSize(), mSteps);
+  corruptInput.set_size(rbmPolicy.VisibleLayer().InSize(), 1);
+  gibbsTemporary.zeros();
+  corruptInput.zeros();
+  negativeSamples.zeros();
+  reset = true;
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+template<typename OptimizerType>
+void RBM<InitializationRuleType, RBMPolicy>::
+    Train(const arma::mat& predictors, OptimizerType& optimizer)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+
+  if (!reset)
+    Reset();
+
+  // Train the model.
+  Timer::Start("rbm_optimization");
+  optimizer.Optimize(*this, parameter);
+  Timer::Stop("rbm_optimization");
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>
+    ::FreeEnergy(arma::mat&& input)
+{
+  return rbmPolicy.FreeEnergy(std::move(input));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>
+    ::Evaluate(const arma::mat& /* parameters*/, const size_t i)
+{
+  corruptInput = arma::round(predictors.col(i));
+  if (!useMonitoringCost)
+  {
+    // Do not use unsafe col predictor as we change the input
+    Gibbs(std::move(predictors.col(i)), std::move(negativeSamples));
+    return (FreeEnergy(std::move(predictors.col(i))) -
+        FreeEnergy(std::move(negativeSamples)));
+  }
+  else
+  {
+    if (persistence)
+    {
+      size_t idx = RandInt(0, predictors.n_rows);
+      corruptInput.row(idx) = 1 - corruptInput.row(idx);
+      return std::log(LogisticFunction::Fn(FreeEnergy(std::move(corruptInput)) -
+          FreeEnergy(std::move(arma::round(predictors.col(i)))))) *
```

Comment:
Well i think the Pseudo likelihood loss should be valid since it is independent of the RBM(depends on cd-k algo). I have to see how to make the cross entropy loss work since ssRBM doesn't have a ForwardFunction.